File: src/apersci/service/allocateInventory.go
package main

import (
	"apersci/input"
	"apersci/output"
	"apersci/soap"
	"database/sql"
	"errors"
	"net/http"
)

func allocateInventory(w http.ResponseWriter, r *http.Request) (err error) {
	req, err := input.AllocateInventoryRequest(r.Body)
	if err != nil {
		return
	}

	tx, err := getDBTransaction()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	isPossible, err := isAllocationPossible(tx, req)
	if err != nil {
		return err
	}
	if !isPossible {
		return errors.New("All of the specified items were not available for allocation.")
	}

	err = allocateAllItems(tx, req)
	if err != nil {
		return
	}

	err = output.AllocateInventoryResponse(w, "")
	if err != nil {
		return
	}

	err = tx.Commit()
	if err != nil {
		return err
	}

	return
}

func isAllocationPossible(tx *sql.Tx, req soap.UpdateRequest) (isPossible bool, err error) {
	isPossible = true
	for _, item := range req.Items {
		// FIXME remove possible SQL injection
		whereClause_UPCorSKU := ""
		if item.PartNumber != "" {
			whereClause_UPCorSKU += " AND fp.SKU = '" + item.PartNumber + "' "
		}
		if item.UPC != "" {
			whereClause_UPCorSKU += " AND fp.UPC = '" + item.UPC + "' "
		}

		rows, err := tx.Query(`
			SELECT (SUM(bp.OnHand) - SUM(bp.Allocated) - lp.SafetyStock) AS Available
			FROM BinProducts bp
				JOIN FulfillerProducts fp ON (fp.FulfillerID = bp.FulfillerID AND fp.SKU = bp.SKU)
				JOIN LocationProducts lp ON (lp.FulfillerID = bp.FulfillerID and lp.SKU = bp.SKU)
				JOIN Bins b ON (b.ID = bp.BinID)
				JOIN Locations l ON (l.ID = b.LocationID)
				JOIN Products p ON (p.UPC = fp.UPC)
			WHERE l.ExternalFulfillerID = $1`+whereClause_UPCorSKU+`
			GROUP BY lp.SafetyStock`,
			item.ExternalLocationID)
		if err != nil {
			return isPossible, err
		}

		adjustedAvailable, err := readIntAndCloseRows(rows)
		if err != nil {
			return isPossible, err
		}

		// check for available amount
		if adjustedAvailable < item.Quantity {
			isPossible = false
			break
		}
	}

	return
}

func allocateAllItems(tx *sql.Tx, req soap.UpdateRequest) (err error) {
	for _, item := range req.Items {
		// FIXME remove possible SQL injection
		whereClause_UPCorSKU := ""
		if item.PartNumber != "" {
			whereClause_UPCorSKU += " AND fp.SKU = '" + item.PartNumber + "' "
		}
		if item.UPC != "" {
			whereClause_UPCorSKU += " AND fp.UPC = '" + item.UPC + "' "
		}

		// collect all bin IDs
		rows, err := tx.Query(`
			SELECT DISTINCT b.ID
			FROM BinProducts bp
				JOIN FulfillerProducts fp ON (fp.FulfillerId = bp.FulfillerID AND fp.SKU = bp.SKU)
				JOIN Bins b ON (b.ID = bp.BinID)
				JOIN Locations l ON (l.ID = b.LocationID)
				JOIN Products p ON (p.UPC = fp.UPC)
			WHERE l.ExternalFulfillerID = $1`+whereClause_UPCorSKU,
			item.ExternalLocationID)
		if err != nil {
			return err
		}

		binIDs, err := readIntSliceAndCloseRows(rows)
		if err != nil {
			return err
		}

		// loop over the bins with the item
		quantityToAllocate := item.Quantity
		for i, binID := range binIDs {
			// determine maximum amount to allocate
			rows, err = tx.Query(`
				SELECT (OnHand - Allocated) AS MaxToAllocate
				FROM BinProducts
				WHERE BinID = $1`,
				binID)

			maxToAllocate, err := readIntAndCloseRows(rows)
			if err != nil {
				return err
			}

			// this will force allocation in case
			// safety stock is a negative value
			isLastBin := i == len(binIDs)-1

			if maxToAllocate > 0 || isLastBin {
				var toAllocate int
				if maxToAllocate > quantityToAllocate || isLastBin {
					toAllocate = quantityToAllocate
				} else {
					toAllocate = maxToAllocate
				}
				quantityToAllocate = quantityToAllocate - toAllocate

				// allocate
				_, err = tx.Exec(`
					UPDATE BinProducts
					SET Allocated = Allocated + $1
					WHERE BinID = $2`,
					toAllocate, binID)
				if err != nil {
					return err
				}
			}
		}
	}

	return
}

File: final/allocateInventory.py
import MySQLdb

def selectItemBins(fulfiller_id, location_id, sku, db):
    cur = db.cur()

    try:
        cur.execute('SELECT OnHand, Allocated, Name ' +
                    'FROM StoredIn WHERE SKU = %s AND FulfillerId = %s AND FulfillerLocationId = %s',
                    (sku, fulfiller_id, location_id))
        return cur.fetchall()
    except Exception, e:
        print e
        return () 
    finally:
        cur.close()

def itemCanBeAllocated(fulfuller_id, location_id, item, db):
    cur = db.cur()
    onhand = 0

    try:
        cur.execute('SELECT SafetyStockLimit FROM StoredAt WHERE SKU = %s AND FulfillerId = %s AND FulfillerLocationId = %s',
                    (item['SKU'], fulfiller_id, location_id))
        safety = cur.fetchone()

        for bin in selectItemBins(fulfiller_id, location_id, item['SKU'], db):
            onhand = onhand + bin[0] 

        if item['quantity'] <= (onhand - safety):
            return True
    except Exception, e:
        print e
    finally:
        cur.close()

    return False

def allocateFromBin(fulfiller_id, location_id, sku, quantity, name, db):
    cur = db.cursor()

    try:
        cur.execute('UPDATE StoredIn' +
                    'SET OnHand = OnHand - %s, Allocated = Allocated + %s' +
                    'WHERE SKU = %s AND FulfillerId = %s AND FulfillerLocationId = %s AND Name = %s',
                    (quantity, quantity, item['SKU'], fulfiller_id, location_id, name))
    except Exception, e:
        print e
    finally:
        cur.close()

def allocateItem(fulfiller_id, location_id, item, db):
    cur = db.cur()
    quantity = item['quantity']
    bins = selectItemBins(fulfiller_id, location_id, sku, db)

    for bin in bins:
        onhand = bin[0]
        name = bin[2]
        allocation = quantity if onhand >= quantity else onhand

        allocateFromBin(fulfiller_id, location_id, sku, allocation, name, db)
        quantity = quantity - allocation
        
        if quantity == 0:
            return

def allocateInventory(fulfiller_id, catalog_id, location_id, items, db):
    cur = db.cur()

    for item in items:
        if not itemCanBeAllocated(fulfiller_id, location_id, item, db):
            print 'Insufficient available inventory. Aborting...'
            return

    for item in items:
        allocateItem(fulfiller_id, location_id, item, db)


